MultiServer.py

```diff
@@ -634,13 +635,34 @@ def get_hint_cost(self, slot):
             return max(1, int(self.hint_cost * 0.01 * len(self.locations[slot])))
         return 0
 
-    def recheck_hints(self, team: typing.Optional[int] = None, slot: typing.Optional[int] = None):
+    def recheck_hints(self, team: typing.Optional[int] = None, slot: typing.Optional[int] = None,
+                      changed: typing.Optional[typing.Set[team_slot]] = None,
+                      already_updated: typing.Optional[typing.Set[team_slot]] = None) -> None:
+        if not already_updated:
+            already_updated = set()
+        needs_updating = set()
         for hint_team, hint_slot in self.hints:
-            if (team is None or team == hint_team) and (slot is None or slot == hint_slot):
-                self.hints[hint_team, hint_slot] = {
-                    hint.re_check(self, hint_team) for hint in
-                    self.hints[hint_team, hint_slot]
-                }
+            if (hint_team, hint_slot) in already_updated:
+                continue
+            if (team != hint_team and team is not None):
+                continue
+            if (slot != hint_slot and slot is not None):
+                continue
+            new_hints: typing.Set[Hint] = set()
+            for hint in self.hints[hint_team, hint_slot]:
+                new_hint = hint.re_check(self, hint_team)
+                new_hints.add(new_hint)
+                if hint == new_hint:
+                    continue
+                if changed is not None:
+                    changed.add((hint_team,hint_slot))
+                for player in (hint.finding_player, hint.receiving_player):
+                    if slot is not None and slot != player:
+                        needs_updating.add((hint_team, player))
+            self.hints[hint_team, hint_slot] = new_hints
+            already_updated.add((hint_team, hint_slot))
+        for hint_team, hint_slot in needs_updating:
+            self.recheck_hints(hint_team, hint_slot, already_updated, changed)
```

Comment:
This recursion seems pretty clunky / wasteful and makes this function super complicated. Am I right in assuming that this is to update the other player's copy of a hint? If yes, couldn't you just use your new `replace_hint` function for that in a more targeted manner?